vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/AbstractLibjnAccumulateScalarEvaluator.java

```diff
@@ -0,0 +1,170 @@
+                arg.getValue(op);
+                addPairs(tempTvp);
+            }
+            for (TaggedValuePointable key1 : tvps.keySet()) {
+                key1.getValue(stringKey);
+                values = tvps.get(key1);
+                if (values.size() > 1) {
+                    FunctionHelper.removeDuplicates(values);
+                    if (values.size() > 1) {
+                        abvs1.reset();
+                        ab.reset(abvs1);
+                        for (TaggedValuePointable pointable : values) {
+                            ab.addItem(pointable);
+                        }
+                        ab.finish();
+                        ob.addItem(stringKey, abvs1);
+                    } else {
+                        ob.addItem(stringKey, values.get(0));
+                    }
+                } else {
+                    addKeyValue(stringKey, values.get(0));
+                }
+            }
+            ob.finish();
+            result.set(abvs);
+        } catch (IOException e) {
+            throw new SystemException(ErrorCode.SYSE0001, e);
+        } finally {
+            ppool.giveBack(tempTvp);
+            for (TaggedValuePointable key1 : tvps.keySet()) {
+                for (TaggedValuePointable value1 : tvps.get(key1)) {
+                    ppool.giveBack(value1);
+                }
+                ppool.giveBack(key1);
+            }
+        }
+    }
+
+    protected abstract void addKeyValue(UTF8StringPointable stringKey, TaggedValuePointable taggedValuePointable) throws IOException;
+
+    private void addPair(TaggedValuePointable tempTvp) throws IOException, SystemException {
+        TaggedValuePointable tvp = isDuplicateKeys(tempTvp, tvps.keySet());
+        value = ppool.takeOne(TaggedValuePointable.class);
+        tempTvp.getValue(stringKey);
+        op.getValue(stringKey, value);
+        if (tvp == null) {
+            key = ppool.takeOne(TaggedValuePointable.class);
+            key.set(tempTvp);
+            values = new ArrayList<>();
+            values.add(value);
+            tvps.put(key, values);
+        } else {
+            values = tvps.get(tvp);
+            values.add(value);
+            tvps.replace(tvp, values);
+        }
+    }
+
+    private void addPairs(TaggedValuePointable tempTvp) throws IOException, SystemException {
+        op.getKeys(tempTvp);
+        if (tempTvp.getTag() == ValueTag.XS_STRING_TAG) {
+            addPair(tempTvp);
+        } else if (tempTvp.getTag() == ValueTag.SEQUENCE_TAG) {
+            tempTvp.getValue(sp1);
+            for (int j = 0; j < sp1.getEntryCount(); ++j) {
+                key = ppool.takeOne(TaggedValuePointable.class);
+                sp1.getEntry(j, tempTvp);
+                addPair(tempTvp);
+            }
+        }
+    }
+
+    protected TaggedValuePointable isDuplicateKeys(TaggedValuePointable key, Set<TaggedValuePointable> pointables) {
```

Comment:
To get the value using a key from the map we would need the exact key in the map, so I am returning the key in the map to get the value using it.